File: getwowstats.go
package main

/////////////////////////////////getwowstats/////////////////////////////
// getwowstats is a cmd-line utility written in GO that can interact    /
//            with the Blizzard World of Warcraft API.                  /
/////////////////////////////////////////////////////////////////////////
///////////////////Author: Gerald L. Hevener Jr. M.S.////////////////////
/////////////////////////////////getwowstats/////////////////////////////

// Import required packages
import (
	"fmt"
	"os"
    "encoding/json"
	"io/ioutil"
	"net/http"
	flag "github.com/ogier/pflag"
)

// Declare Variable types
var (
	region			      string
	character		      string
	realm			  	  string
	method		       	  string
	version				  string
)

// Declare program's main function
func main() {
	
	// If user does not supply flags print tool usage
	if flag.NFlag() == 0 {
		fmt.Printf("Usage: %s [options]\n", os.Args[0])
		fmt.Println("Options:")
		flag.PrintDefaults()
		os.Exit(1)
	}

	// Uncomment below to debug user supplied cmd line options
	//fmt.Printf("Using API token for World of Warcraft: %s\n", wow_api_token)
	//fmt.Printf("Using two letter country code region : %s\n", region)
    //fmt.Printf("Using character name: %s\n", character)
	//fmt.Printf("Using realm: %s\n", realm)
	//fmt.Printf("Calling method: %s\n", method)

}

// Declare init function to parse user supplied cmd line options
func init() {

    flag.StringVarP(&region, "locale", "l", "", "Your two letter country code ( locale )")
	flag.StringVarP(&character, "character", "c", "", "Name of WoW character to lookup stats about")
	flag.StringVarP(&realm, "realm", "r", "", "Realm of WoW character to lookup stats about")
	flag.StringVarP(&method, "method", "m", "", "Method in getwowstats to call")
	flag.StringVarP(&version, "version", "v", "", "Current version of getwowstats")
   
	// Parse CMD line ARGS
	flag.Parse()

	// Process cmd line ARGS and call appropriate function
    if method == "GetProfileSummary" {
	  GetProfileSummary()
	}

	// Print software version of getwowstats
	if version != "" {
		GetWowStatsVersion()
	}

}

// Set version of getwowstats
func GetWowStatsVersion() {
 
	getwowstats_version := "1.0.0"
	fmt.Println("getwowstats: Version " + getwowstats_version)

}

// Retrieve Character Profile summary
func GetProfileSummary() {

    // Get WoW API token
	val, env_var := os.LookupEnv("WOW_API_TOKEN")
    if env_var {
	  // Uncomment to debug value of ENV VAR WOW_API_TOKEN	
	  //fmt.Println(val)
	  wow_api_endpoint := "https://us.api.blizzard.com/profile/wow/character/" + realm + "/" + character + "?namespace=profile-" + region + "&locale=en_US&access_token=" + val
	
	// Create an HTTP client
    client := &http.Client{}

	// Process API response and errors
    resp, err := client.Get(wow_api_endpoint)
	if err != nil {
		fmt.Println("Failed to call API!", err)
	} else {
		//Print API response
		data, _ := ioutil.ReadAll(resp.Body)

		// Uncomment below to print entire API json response
        // fmt.Println(string(data))

		// Unmarshall to decode API json response
	    var profile map[string]interface{}
		json.Unmarshal([]byte(data),&profile)

		// Print out character profile summary
		fmt.Println("Name:", profile["name"])
		fmt.Println("Gender:", profile["gender"].(map[string]interface{})["name"])
		fmt.Println("Faction:", profile["faction"].(map[string]interface{})["name"])
		fmt.Println("Race:", profile["race"].(map[string]interface{})["name"])
		fmt.Println("Class:", profile["character_class"].(map[string]interface{})["name"])
		fmt.Println("Realm:", profile["realm"].(map[string]interface{})["name"])
		fmt.Println("Level:", profile["level"])
		fmt.Println("Active Spec:", profile["active_spec"].(map[string]interface{})["name"])
		fmt.Println("Experience:", profile["experience"])
        fmt.Println("Achieveent Points:", profile["achievement_points"])
        fmt.Println("Last Login:", profile["last_login_timestamp"])
        fmt.Println("Avg. Item Lvl.:", profile["average_item_level"])
        fmt.Println("Equipped Item Lvl.:", profile["equipped_item_level"])

		// Close API connection
		defer resp.Body.Close()
	}   
	
	} 
}	

File: src/bin/getwowstats.rs
// Import the getwowstats library to use in our CLI tool.
use clap::Parser;
use std::collections::HashMap;
use std::env;

#[derive(Parser, Debug)]
#[clap(author, version, about, long_about = None)]

struct Args {
    /// World of Warcraft character name
    #[clap(short, long)]
    character: String,

    /// Your two letter country code ( e.g. locale )
    #[clap(short, long)]
    locale: String,

    /// Method in getwowstats to call
    #[clap(short, long)]
    method: String,

    /// Realm of WoW character to lookup stats about
    #[clap(short, long)]
    //counter: u8,
    realm: String,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();
    
    println!("Character {}", args.character);
    println!("Locale {}", args.locale);
    println!("Method {}", args.method);
    println!("Realm {}", args.realm);

    // Declare variables
    //base_api_url = "https://us.api.blizzard.com/profile/wow/character/";
    //forward_slash = "/";
    //let wow_api_endpoint = format!("{}{}{}{}{}{}{}", base_api_url, args.realm, forward_slash, args.character, "?namespace=profile-", args.region, "&locale=en_US&access_token=");
    
    let wow_api_token = "WOW_API_TOKEN";
    let wow_token = env::var("WOW_API_TOKEN").unwrap_or("none".to_string());
    println!("WOW_API_TOKEN is set to {}", wow_api_token);

    // Declare local variables
    //let wow_api_token = "abc123";
    let base_api_url = "https://us.api.blizzard.com/profile/wow/character/".to_owned();
    let base_api_url2 = base_api_url + &args.realm.to_owned();
    let base_api_url3 = base_api_url2 + &"/".to_owned();
    let base_api_url4 = base_api_url3 + &args.character.to_owned();
    let base_api_url5 = base_api_url4 + &"?namespace=profile-".to_owned();
    let base_api_url6 = base_api_url5 + &args.locale.to_owned();
    let base_api_url7 = base_api_url6 + &"&locale=en_US&access_token=".to_owned();
    let wow_api_url = base_api_url7 + &wow_token.to_owned();

    println!("{}", wow_api_url);

    //let resp = reqwest::get("https://httpbin.org/ip")
    let resp = reqwest::get(wow_api_url)
        .await?
        .json::<HashMap<String, String>>()
        .await?;
    println!("{:#?}", resp);
    Ok(())

}
